In [1]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px

In [2]:
os.chdir('/Users/omarespinoza/Documents/PUENTE_DATA/Materiales San Pedro')


In [34]:
ruta_archivo = "RepArticulosCom.xlsx"
df = pd.read_excel(ruta_archivo,header=7)

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/openpyxl/styles/stylesheet.py:237: UserWarning: Workbook contains no default style, apply openpyxl's default
  warn("Workbook contains no default style, apply openpyxl's default")


In [35]:
df.head(10)

,Clave,Unnamed: 1,Descripción,Unnamed: 3,Unnamed: 4,Departamento,Unnamed: 6,Unnamed: 7,Unnamed: 8,Categoría,...,Unnamed: 13,Factor,Unnamed: 15,Precio,Unnamed: 17,Unnamed: 18,Unnamed: 19,Importe,Unnamed: 21,Unnamed: 22
0,15819,NaN,1 banda de lija 3 x 21' p/madera grano 120 en ...,NaN,NaN,Truper,NaN,NaN,NaN,Accesorios máquinas,...,NaN,1.0,NaN,$ 22.01,NaN,NaN,NaN,$ 110.05,NaN,NaN
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,17771,NaN,1 banda de lija 3 x 21' p/vidrio grano 400 en ...,NaN,NaN,Truper,NaN,NaN,NaN,Accesorios máquinas,...,NaN,1.0,NaN,$ 25.00,NaN,NaN,NaN,$ 250.00,NaN,NaN
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,68052,NaN,100 bolsas plásticas biodegradables de 20 x 40...,NaN,NaN,Truper,NaN,NaN,NaN,Artículos promocionales,...,NaN,1.0,NaN,$ 49.00,NaN,NaN,NaN,$ 147.00,NaN,NaN
5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,68052,NaN,100 bolsas plásticas biodegradables de 20 x 40...,NaN,NaN,Truper,NaN,NaN,NaN,Artículos promocionales,...,NaN,1.0,NaN,$ 42.24,NaN,NaN,NaN,$ 42.24,NaN,NaN
7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,68053,NaN,100 bolsas plásticas biodegradables de 27 x 50...,NaN,NaN,Truper,NaN,NaN,NaN,Artículos promocionales,...,NaN,1.0,NaN,$ 85.34,NaN,NaN,NaN,$ 85.34,NaN,NaN
9,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [36]:
df.tail(10)

,Clave,Unnamed: 1,Descripción,Unnamed: 3,Unnamed: 4,Departamento,Unnamed: 6,Unnamed: 7,Unnamed: 8,Categoría,...,Unnamed: 13,Factor,Unnamed: 15,Precio,Unnamed: 17,Unnamed: 18,Unnamed: 19,Importe,Unnamed: 21,Unnamed: 22
6084,YPS4,NaN,"YEE PVC SANITARIO DE 4""",NaN,NaN,SANITARIO,NaN,NaN,NaN,PVC SANITARIO,...,NaN,1.0,NaN,$ 41.53,NaN,NaN,NaN,$ 872.13,NaN,NaN
6085,YPS4,NaN,"YEE PVC SANITARIO DE 4""",NaN,NaN,SANITARIO,NaN,NaN,NaN,PVC SANITARIO,...,NaN,1.0,NaN,$ 40.00,NaN,NaN,NaN,"$ 2,120.00",NaN,NaN
6086,YPS4,NaN,"YEE PVC SANITARIO DE 4""",NaN,NaN,SANITARIO,NaN,NaN,NaN,PVC SANITARIO,...,NaN,1.0,NaN,$ 46.08,NaN,NaN,NaN,$ 552.96,NaN,NaN
6087,YPS42,NaN,"YEE PVC SANITARIO DE 4"" X 2""",NaN,NaN,SANITARIO,NaN,NaN,NaN,PVC SANITARIO,...,NaN,1.0,NaN,$ 26.00,NaN,NaN,NaN,$ 312.00,NaN,NaN
6088,0070,NaN,YESO MAXIMO 40KG,NaN,NaN,POLVOS,NaN,NaN,NaN,PETREOS Y CONGLOMERANTES,...,NaN,1.0,NaN,$ 107.00,NaN,NaN,NaN,"$ 10,700.00",NaN,NaN
6089,0070,NaN,YESO MAXIMO 40KG,NaN,NaN,POLVOS,NaN,NaN,NaN,PETREOS Y CONGLOMERANTES,...,NaN,1.0,NaN,$ 125.50,NaN,NaN,NaN,"$ 3,137.50",NaN,NaN
6090,18646,NaN,"Zapapico 5 lb con mango de 36', Truper",NaN,NaN,Truper,NaN,NaN,NaN,Zapapicos y talachos nacionales,...,NaN,1.0,NaN,$ 275.00,NaN,NaN,NaN,"$ 1,100.00",NaN,NaN
6091,18646,NaN,"Zapapico 5 lb, mango fibra de vidrio 36', Truper",NaN,NaN,Truper,NaN,NaN,NaN,Zapapicos y talachos nacionales,...,NaN,1.0,NaN,$ 275.00,NaN,NaN,NaN,"$ 1,100.00",NaN,NaN
6092,46310,NaN,"ZI-C Blíster con 2 pilas zinc-carbón C, VOLTECK",NaN,NaN,Volteck,NaN,NaN,NaN,Baterías,...,NaN,1.0,NaN,$ 12.93,NaN,NaN,NaN,$ 77.58,NaN,NaN
6093,102967,NaN,"ZP-5MM Zapapico 5 lb, mango de madera 36', Truper",NaN,NaN,Truper,NaN,NaN,NaN,Zapapicos y talachos nacionales,...,NaN,1.0,NaN,$ 257.76,NaN,NaN,NaN,"$ 2,062.08",NaN,NaN


In [37]:
columnas_vacias = df.columns[df.isna().all()].tolist()

In [38]:
print(f"Se encontraron {len(columnas_vacias)} columnas completamente vacías.")
if columnas_vacias:
    print("Columnas detectadas para eliminación:", columnas_vacias)
else:
    print("No hay columnas puras de NaN en este DataFrame.")

Se encontraron 15 columnas completamente vacías.
Columnas detectadas para eliminación: ['Unnamed: 1', 'Unnamed: 3', 'Unnamed: 4', 'Unnamed: 6', 'Unnamed: 7', 'Unnamed: 8', 'Unnamed: 10', 'Unnamed: 11', 'Unnamed: 13', 'Unnamed: 15', 'Unnamed: 17', 'Unnamed: 18', 'Unnamed: 19', 'Unnamed: 21', 'Unnamed: 22']


In [39]:
if columnas_vacias:
    df = df.drop(columns=columnas_vacias)
    print(f"¡Limpieza exitosa!")

¡Limpieza exitosa!


In [40]:
df.head(10)

,Clave,Descripción,Departamento,Categoría,Cant,Factor,Precio,Importe
0,15819,1 banda de lija 3 x 21' p/madera grano 120 en ...,Truper,Accesorios máquinas,5.0,1.0,$ 22.01,$ 110.05
1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,17771,1 banda de lija 3 x 21' p/vidrio grano 400 en ...,Truper,Accesorios máquinas,10.0,1.0,$ 25.00,$ 250.00
3,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,3.0,1.0,$ 49.00,$ 147.00
5,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,1.0,1.0,$ 42.24,$ 42.24
7,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
8,68053,100 bolsas plásticas biodegradables de 27 x 50...,Truper,Artículos promocionales,1.0,1.0,$ 85.34,$ 85.34
9,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [41]:
df = df.dropna(how='all')

In [42]:
df.head(10)

,Clave,Descripción,Departamento,Categoría,Cant,Factor,Precio,Importe
0,15819,1 banda de lija 3 x 21' p/madera grano 120 en ...,Truper,Accesorios máquinas,5.0,1.0,$ 22.01,$ 110.05
2,17771,1 banda de lija 3 x 21' p/vidrio grano 400 en ...,Truper,Accesorios máquinas,10.0,1.0,$ 25.00,$ 250.00
4,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,3.0,1.0,$ 49.00,$ 147.00
6,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,1.0,1.0,$ 42.24,$ 42.24
8,68053,100 bolsas plásticas biodegradables de 27 x 50...,Truper,Artículos promocionales,1.0,1.0,$ 85.34,$ 85.34
10,68053,100 bolsas plásticas biodegradables de 27 x 50...,Truper,Artículos promocionales,4.0,1.0,$ 98.99,$ 395.96
12,68054,100 bolsas plásticas biodegradables de 34 x 60...,Truper,Artículos promocionales,4.0,1.0,$ 149.00,$ 596.00
14,68054,100 bolsas plásticas biodegradables de 34 x 60...,Truper,Artículos promocionales,1.0,1.0,$ 128.45,$ 128.45
16,46123,"2 Contactos de PVC, Volteck",Volteck,Clavijas y Contactos,10.0,1.0,$ 17.01,$ 170.10
17,C20101,201 COLOR P/ CEMENTO AZUL REY 1KG,CPP,POLVOS Y PIGMENTOS,6.0,1.0,$ 48.02,$ 288.12


In [43]:
df.info()

<class 'pandas.DataFrame'>
Index: 4152 entries, 0 to 6093
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0     Clave       4136 non-null   str    
 1   Descripción   4152 non-null   str    
 2   Departamento  4136 non-null   str    
 3   Categoría     4136 non-null   str    
 4   Cant          4136 non-null   float64
 5   Factor        4136 non-null   float64
 6   Precio        4136 non-null   str    
 7   Importe       4136 non-null   str    
dtypes: float64(2), str(6)
memory usage: 655.9 KB


In [44]:
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

In [45]:
df.rename(columns={'descripción': 'descripcion'}, inplace=True)
df.rename(columns={'cant': 'cantidad_producto_compra'}, inplace=True)
df.rename(columns={'precio': 'precio_unidad_compra'}, inplace=True)
df.rename(columns={'importe': 'importe_total_compra'}, inplace=True)
df.rename(columns={'categoría': 'categoria'}, inplace=True)



In [46]:
df['precio_unidad_compra'] = df['precio_unidad_compra'].str.replace('$', '', regex=False)
df['importe_total_compra'] = df['importe_total_compra'].str.replace('$', '', regex=False)


In [47]:
df['precio_unidad_compra'] = pd.to_numeric(df['precio_unidad_compra'], errors='coerce')
df['importe_total_compra'] = pd.to_numeric(df['importe_total_compra'], errors='coerce')


In [48]:
print('Existen ',df.isnull().sum().sum(),' valores no existentes')

Existen  1541  valores no existentes


In [49]:
missing_values_count = df.isnull().sum()
print ('Columnas que tienen nulos')
print('-------------------------')
missing_values_count

Columnas que tienen nulos
-------------------------


clave                         16
descripcion                    0
departamento                  16
categoria                     16
cantidad_producto_compra      16
factor                        16
precio_unidad_compra         170
importe_total_compra        1291
dtype: int64

In [50]:
df_nulos = df[df.isnull().any(axis=1)]


In [51]:
df_nulos

,clave,descripcion,departamento,categoria,cantidad_producto_compra,factor,precio_unidad_compra,importe_total_compra
21,C20801,208 COLOR P/ CEMENTO AMARILLO OXIDO 1KG,CPP,POLVOS Y PIGMENTOS,48.0,1.0,47.33,NaN
23,V400C,400 VINILICA MURAL BLANCO 19 L,CPP,VINILICA,8.0,1.0,NaN,NaN
24,V400G,400 VINILICA MURAL BLANCO 4 LT,CPP,VINILICA,20.0,1.0,241.51,NaN
30,1477,423 VINILICA MURAL ALMENDRA 19L,SIN DEFINIR,SIN DEFINIR,4.0,1.0,NaN,NaN
66,PAQBAÑO,ACCESORIOS DE INSTALACION PARA BAÑO,SIN DEFINIR,SIN DEFINIR,2.0,1.0,NaN,NaN
...,...,...,...,...,...,...,...,...
6088,0070,YESO MAXIMO 40KG,POLVOS,PETREOS Y CONGLOMERANTES,100.0,1.0,107.00,NaN
6089,0070,YESO MAXIMO 40KG,POLVOS,PETREOS Y CONGLOMERANTES,25.0,1.0,125.50,NaN
6090,18646,"Zapapico 5 lb con mango de 36', Truper",Truper,Zapapicos y talachos nacionales,4.0,1.0,275.00,NaN
6091,18646,"Zapapico 5 lb, mango fibra de vidrio 36', Truper",Truper,Zapapicos y talachos nacionales,4.0,1.0,275.00,NaN


TENEMOS 16 PRODUCTOS QUE NUNCA SE HAN COMPRADO 

In [52]:
df_id_nulos = df[df['cantidad_producto_compra'].isnull()]
df_id_nulos

,clave,descripcion,departamento,categoria,cantidad_producto_compra,factor,precio_unidad_compra,importe_total_compra
290,NaN,"110, FIERO",NaN,NaN,NaN,NaN,NaN,NaN
871,NaN,Pretul,NaN,NaN,NaN,NaN,NaN,NaN
957,NaN,"AWG blanco, Volteck",NaN,NaN,NaN,NaN,NaN,NaN
1862,NaN,127V Truper,NaN,NaN,NaN,NaN,NaN,NaN
1899,NaN,polos+tierra,NaN,NaN,NaN,NaN,NaN,NaN
2671,NaN,"corto,hexagonal, BASIC",NaN,NaN,NaN,NaN,NaN,NaN
2760,NaN,con 24 piezas,NaN,NaN,NaN,NaN,NaN,NaN
3567,NaN,"gabinete, Volteck",NaN,NaN,NaN,NaN,NaN,NaN
4279,NaN,"presión, tubo 50ml, FOSET",NaN,NaN,NaN,NaN,NaN,NaN
4339,NaN,"Grip, Pretul",NaN,NaN,NaN,NaN,NaN,NaN


In [53]:
df_id_nulos_importe_total = df[df['importe_total_compra'].isnull()]
df_id_nulos_importe_total

,clave,descripcion,departamento,categoria,cantidad_producto_compra,factor,precio_unidad_compra,importe_total_compra
21,C20801,208 COLOR P/ CEMENTO AMARILLO OXIDO 1KG,CPP,POLVOS Y PIGMENTOS,48.0,1.0,47.33,NaN
23,V400C,400 VINILICA MURAL BLANCO 19 L,CPP,VINILICA,8.0,1.0,NaN,NaN
24,V400G,400 VINILICA MURAL BLANCO 4 LT,CPP,VINILICA,20.0,1.0,241.51,NaN
30,1477,423 VINILICA MURAL ALMENDRA 19L,SIN DEFINIR,SIN DEFINIR,4.0,1.0,NaN,NaN
66,PAQBAÑO,ACCESORIOS DE INSTALACION PARA BAÑO,SIN DEFINIR,SIN DEFINIR,2.0,1.0,NaN,NaN
...,...,...,...,...,...,...,...,...
6088,0070,YESO MAXIMO 40KG,POLVOS,PETREOS Y CONGLOMERANTES,100.0,1.0,107.00,NaN
6089,0070,YESO MAXIMO 40KG,POLVOS,PETREOS Y CONGLOMERANTES,25.0,1.0,125.50,NaN
6090,18646,"Zapapico 5 lb con mango de 36', Truper",Truper,Zapapicos y talachos nacionales,4.0,1.0,275.00,NaN
6091,18646,"Zapapico 5 lb, mango fibra de vidrio 36', Truper",Truper,Zapapicos y talachos nacionales,4.0,1.0,275.00,NaN


TENEMOS 1291 PRODCUTOS QUE TIENEN CANTIDAD DE PRODUCTO COMPRADO PERO NO TIENEN IMPORTE TOTAL, INCLUSO TIENEN EL PRECIO POR UNIDAD



In [54]:
df_id_nulos_precio_unidad = df[df['precio_unidad_compra'].isnull()]
df_id_nulos_precio_unidad

,clave,descripcion,departamento,categoria,cantidad_producto_compra,factor,precio_unidad_compra,importe_total_compra
23,V400C,400 VINILICA MURAL BLANCO 19 L,CPP,VINILICA,8.0,1.0,NaN,NaN
30,1477,423 VINILICA MURAL ALMENDRA 19L,SIN DEFINIR,SIN DEFINIR,4.0,1.0,NaN,NaN
66,PAQBAÑO,ACCESORIOS DE INSTALACION PARA BAÑO,SIN DEFINIR,SIN DEFINIR,2.0,1.0,NaN,NaN
290,NaN,"110, FIERO",NaN,NaN,NaN,NaN,NaN,NaN
388,12092,"Aspiradora de sólidos y líquidos de 6 gal, plá...",Truper,Aspiradoras,1.0,1.0,NaN,NaN
...,...,...,...,...,...,...,...,...
5902,V429C,V429 VINILICA MURAL CAFÉ CORDOBA 19 L,CPP,VINILICA,2.0,1.0,NaN,NaN
5906,V431C,V431 VINILICA MURAL PERLA 19 L,CPP,VINILICA,3.0,1.0,NaN,NaN
5912,NaN,Foset,NaN,NaN,NaN,NaN,NaN,NaN
6040,NaN,"kg, Foset",NaN,NaN,NaN,NaN,NaN,NaN


In [55]:
df = df.dropna(subset=['cantidad_producto_compra'])

In [56]:
missing_values_count = df.isnull().sum()
print ('Columnas que tienen nulos')
print('-------------------------')
missing_values_count

Columnas que tienen nulos
-------------------------


clave                          0
descripcion                    0
departamento                   0
categoria                      0
cantidad_producto_compra       0
factor                         0
precio_unidad_compra         154
importe_total_compra        1275
dtype: int64

In [57]:
df=df.fillna(0)

In [58]:
missing_values_count = df.isnull().sum()
print ('Columnas que tienen nulos')
print('-------------------------')
missing_values_count

Columnas que tienen nulos
-------------------------


clave                       0
descripcion                 0
departamento                0
categoria                   0
cantidad_producto_compra    0
factor                      0
precio_unidad_compra        0
importe_total_compra        0
dtype: int64

In [59]:
df.head(10)

,clave,descripcion,departamento,categoria,cantidad_producto_compra,factor,precio_unidad_compra,importe_total_compra
0,15819,1 banda de lija 3 x 21' p/madera grano 120 en ...,Truper,Accesorios máquinas,5.0,1.0,22.01,110.05
2,17771,1 banda de lija 3 x 21' p/vidrio grano 400 en ...,Truper,Accesorios máquinas,10.0,1.0,25.00,250.00
4,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,3.0,1.0,49.00,147.00
6,68052,100 bolsas plásticas biodegradables de 20 x 40...,Truper,Artículos promocionales,1.0,1.0,42.24,42.24
8,68053,100 bolsas plásticas biodegradables de 27 x 50...,Truper,Artículos promocionales,1.0,1.0,85.34,85.34
10,68053,100 bolsas plásticas biodegradables de 27 x 50...,Truper,Artículos promocionales,4.0,1.0,98.99,395.96
12,68054,100 bolsas plásticas biodegradables de 34 x 60...,Truper,Artículos promocionales,4.0,1.0,149.00,596.00
14,68054,100 bolsas plásticas biodegradables de 34 x 60...,Truper,Artículos promocionales,1.0,1.0,128.45,128.45
16,46123,"2 Contactos de PVC, Volteck",Volteck,Clavijas y Contactos,10.0,1.0,17.01,170.10
17,C20101,201 COLOR P/ CEMENTO AZUL REY 1KG,CPP,POLVOS Y PIGMENTOS,6.0,1.0,48.02,288.12


In [60]:
df.info()

<class 'pandas.DataFrame'>
Index: 4136 entries, 0 to 6093
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   clave                     4136 non-null   str    
 1   descripcion               4136 non-null   str    
 2   departamento              4136 non-null   str    
 3   categoria                 4136 non-null   str    
 4   cantidad_producto_compra  4136 non-null   float64
 5   factor                    4136 non-null   float64
 6   precio_unidad_compra      4136 non-null   float64
 7   importe_total_compra      4136 non-null   float64
dtypes: float64(4), str(4)
memory usage: 589.0 KB


In [61]:
import unicodedata
def limpiar_nombre(nombre):
    nombre = nombre.lower()
    nombre = unicodedata.normalize('NFKD', nombre).encode('ASCII', 'ignore').decode('utf-8')
    nombre = nombre.strip().replace(' ', '_')
    return nombre


In [62]:
df.columns = [limpiar_nombre(col) for col in df.columns]

In [63]:
df.info()

<class 'pandas.DataFrame'>
Index: 4136 entries, 0 to 6093
Data columns (total 8 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   clave                     4136 non-null   str    
 1   descripcion               4136 non-null   str    
 2   departamento              4136 non-null   str    
 3   categoria                 4136 non-null   str    
 4   cantidad_producto_compra  4136 non-null   float64
 5   factor                    4136 non-null   float64
 6   precio_unidad_compra      4136 non-null   float64
 7   importe_total_compra      4136 non-null   float64
dtypes: float64(4), str(4)
memory usage: 589.0 KB
